### ISSUE TITLE
Dynamic Respawn Timers

### ISSUE THREAD

**Issue by ccshiro** [ccshiro/cc-buglist#1067]
Both Game Objects and NPCs need to have dynamic respawn timers. In other words, if an area have NPCs dying quickly, or Game Objects looted in rapid fashion, it needs to churn them out at an increased rate.
#1066 needs to be implemented before this item.

**Comment by KevinTyrrell** [comment]
Is this a quality of life change? For example, on retail I thought that if you threw 10,000 players in one small area and had them kill all the mobs and loot objects, the mobs would still spawn at their usual rates. Is this change made to prevent the masses from complaining about slow respawn timers? I'm confused.

**Comment by ccshiro** [comment]
@KevinTyrrell no, this is not a QoL change. This is a mechanic that existed in retail as well (at least starting TBC, not sure if it was in vanilla).

**Comment by Krack3n77** [comment]
@ccshiro I'm almost positive this was implemented shortly after the launch of TBC. Wish I could find the first time Blizz mentioned this as a thing. They were planning for it after the opening of AQ after all the lessons they learned from that.

**Comment by Ickler** [comment]
I distinctly remember HFP the first day of TBC, it was having massively increased rates of spawn for objects and NPC's alike. 

And I think this one is a huuuge game changer upon launch, and how fast people will be able to move into other areas :) Glad to see this is a priority

**Comment by ghost** [comment]
@ccHudsonHawk I can only find people referencing Blizzard implementing this which leads me to believe they discussed it on the old forums, before the days of the magnificent mmo-champion blue's tracker. Then again, I can't even find people quoting any such post or article. They always just say Blizz is implementing it with BC. 

I think everyone accepts that it was a thing added with BC though, especially if you played through that launch and _tried_ to do the quests in Zeth'Gor by yourself. Kill a warlock, sit down to drink, warlock respawns a second later.

**Comment by KimmoKM** [comment]
A Wowhead [comment](http://www.wowhead.com/npc=21060/enraged-air-spirit#comments:id=118865) from 2.1.3 mentions farmed quest mobs respawning at a rapid rate with 9-10 people farming them, stating that their status as quest mobs is responsible for this effect.  I'm not convinced about quest mobs being handled any differently, but he certainly seems convinced competition in the area is actually a good thing rather than a handicap (as you might expect if there was no dynamic spawn rates).

**Comment by Zindai** [comment]
The current rates seem a little off. I'm the only person in Durotar, and clearing a cave **alone** at a reasonable pace gives me respawns aggroing me from behind. The respawn rate should most likely be tied to the number of people nearby as well, not _only_ kill speed, or at the very least be reduced a little. The same goes for quest objects. I had the same issue in Hellfire Peninsula where even an open area gets crowded with enemies real fast, both when you're killing them on your own and with a group - a mana break is your doom!

To sum it up: The current fast rates make it hard to get deep into an area for a specific object or named mob without getting screwed over by respawns.

**Comment by tedrasiel** [comment]
I just want to say that i share the opinion of Zindai here. It REALLY should be a combination of the amount of people in a certain area + the killrate of mobs in the same given area.

**Comment by yanech** [comment]
One way to do it :

if a spot has 40 quest mobs, and the living mobs number goes under a threshold (less than 20 for instance), it respans mobs up to the threshold.
If the living mobs number stays above the threshold, mobs are respaned by a timer.

I think this is an effective and easy way to code it.

If you know how it was on retail, or have a better idea, then discard my comment.

Sorry for my bad english.

**Comment by Zindai** [comment]
@yanech
I like your idea, but it can go very wrong if the minimum number is set too high. 2-3 people or groups questing in a cave at the same time could quickly wipe each other out by forcing premature respawns. Sadly I speak from experience.
But of course this is something to keep in mind regardless of the way you choose to speed up respawns.

**Comment by yanech** [comment]
How does agro work on retail ?
Are you attacked as soon as the mob respawn on you, or do you have sonething like 5/10 sec before agroing if the mob is left alone ?

**Comment by Zindai** [comment]
You have a couple of seconds to get away.

This was also the case on the server where I had my unfortunate experience, but due to the high minimum there was nowhere safe to run.

**Comment by yanech** [comment]
A unique variable should be used world wide (30% ?? or something else), so it can be quickly modified.
In TBC, I don't see where this should be a problem.

You can set a minimal respan time too (20s ? 30s ? more ?)

But before thinking further, shiro should say if this interrests him.

I did some AoE grinding on retail in HF Peninsula on vultures or orcs.
I don't remember having a very fast respan time.
